syntax = "proto3";

package gym_mcp_server;

// Gym Environment Service - exposes gym operations via gRPC
service GymService {
  // Reset the environment to its initial state
  rpc Reset(ResetRequest) returns (ResetResponse);
  
  // Take an action in the environment
  rpc Step(StepRequest) returns (StepResponse);
  
  // Render the current state of the environment
  rpc Render(RenderRequest) returns (RenderResponse);
  
  // Get information about the environment
  rpc GetInfo(GetInfoRequest) returns (GetInfoResponse);
  
  // Close the environment and free resources
  rpc Close(CloseRequest) returns (CloseResponse);
  
  // Health check endpoint
  rpc Health(HealthRequest) returns (HealthResponse);
}

// Request/Response messages

message ResetRequest {
  optional int64 seed = 1;  // Random seed for reproducible episodes
}

message ResetResponse {
  bool success = 1;
  optional string error = 2;
  Observation observation = 3;
  map<string, string> info = 4;  // JSON-encoded info dict
  bool done = 5;
}

message StepRequest {
  Action action = 1;  // The action to take
}

message StepResponse {
  bool success = 1;
  optional string error = 2;
  Observation observation = 3;
  double reward = 4;
  bool done = 5;
  bool truncated = 6;
  map<string, string> info = 7;  // JSON-encoded info dict
}

message RenderRequest {
  optional string mode = 1;  // Render mode (e.g., rgb_array, human)
}

message RenderResponse {
  bool success = 1;
  optional string error = 2;
  optional string render = 3;  // Base64-encoded image or JSON-encoded data
  optional string mode = 4;
  optional string type = 5;  // Type: image_base64, array, text, string
  repeated int64 shape = 6;  // Shape of render output (for arrays/images)
}

message GetInfoRequest {
  // Empty request
}

message GetInfoResponse {
  bool success = 1;
  optional string error = 2;
  EnvInfo env_info = 3;
}

message CloseRequest {
  // Empty request
}

message CloseResponse {
  bool success = 1;
  optional string error = 2;
  string status = 3;  // "closed" or "error"
}

message HealthRequest {
  // Empty request
}

message HealthResponse {
  string status = 1;  // "healthy"
  string env_id = 2;
  optional string render_mode = 3;
}

// Nested message types

message Observation {
  oneof value {
    ScalarObservation scalar = 1;
    ArrayObservation array = 2;
    DictObservation dict = 3;
  }
}

message ScalarObservation {
  oneof value {
    int64 int_value = 1;
    double float_value = 2;
    bool bool_value = 3;
    string string_value = 4;  // Fallback for complex types (JSON-encoded)
  }
}

message ArrayObservation {
  repeated double values = 1;  // Flattened array
  repeated int64 shape = 2;
}

message DictObservation {
  map<string, Observation> values = 1;
}

message Action {
  oneof value {
    int64 int_value = 1;
    double float_value = 2;
    ArrayAction array_value = 3;
    DictAction dict_value = 4;
  }
}

message ArrayAction {
  repeated double values = 1;  // Flattened array
  repeated int64 shape = 2;
}

message DictAction {
  map<string, Action> values = 1;
}

message EnvInfo {
  optional string id = 1;
  string action_space = 2;
  string observation_space = 3;
  optional RewardRange reward_range = 4;
  optional int64 action_space_size = 5;
  repeated int64 action_space_shape = 6;
  repeated int64 observation_space_shape = 7;
  map<string, string> observation_space_spaces = 8;
}

message RewardRange {
  double min = 1;
  double max = 2;
}

